### PR TITLE
New collection iteration

### PR DESCRIPTION
**Pull request by devmanhinton**
@yaymukund cc @bglusman 

This commit changes collection reactions' syntax & name.

Collection reactions are now called iteration reactions and use
data-each, as opposed to a parent with a plural name of their own class name.

Use:

HTML:

``` html
<div data-each>
  <div>
      Hello my name is <span class='firstName-'></span>
  </div>
</div>
```

JS:

``` javascript
var collection = new Backbone.Collection([new Backbone.Model({firstName: "Brian"}),
                       new Backbone.Model({firstName: "Devon"}),
                       new Backbone.Model({firstName: "Mukund"}) ])

EndDash.getTemplate('hello', collection)
```

Output:

```
Brian
Devon
Mukund
```

We have another branch new-collection-iteration-deprecation which has these changes & backwards compatibly with collection reactions.

### CODE REVIEW

**Comment by yaymukund** [comment]
Why did you change `collection` to `iteration`?

File: lib/reactions/iteration.js [hunk 1]
```diff
@@ -0,0 +1,118 @@
+var Parser = require("../parser")
+  , TemplateStore = require('../template_store')
```
Comment (yaymukund) [inline lib/reactions/iteration.js]: Don't start lines with commas.

File: lib/reactions/iteration.js [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  , rules = require("../rules");
+
+var IterationReaction = Reaction.extend({
+
+  addAll: function(models, index) {
+    index = index || 0;
+    if(_.isArray(models)) {
+      models = _(models);
+    }; //Hit only if array is passed in and iterated without scoping
+    models.each(function(item, index) {
```
Comment (yaymukund) [inline lib/reactions/iteration.js]: `models.each(this.add, this)`?

File: lib/reactions/iteration.js [hunk 3]
```diff
@@ -0,0 +1,118 @@
+var Parser = require("../parser")
```
Comment (yaymukund) [inline lib/reactions/iteration.js]: I think the name change should be its own commit because it makes this really hard to code review. Is there some way to get a diff of just what has changed in this file?

File: lib/reactions/iteration.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+      , children;
+
+    template.template.data("model", model);
+
+    if(typeof index === "undefined") {
+      this.templates.push(template);
+      this.el.append(template.template);
+    } else if(index === 0) {
+      this.templates.unshift(template);
+      this.el.prepend(template.template);
```
Comment (yaymukund) [inline lib/reactions/iteration.js]: We should be using `template.el` instead of `template.template` in new code.

File: lib/reactions/iteration.js (hunk 1 above)
Comment (devmanhinton) [inline lib/reactions/iteration.js]: I personally like that style better, but I'm flexible and I'll change my style to your preference

Comment (yaymukund) [inline lib/reactions/iteration.js]: Oops, should've been more specific. This style of assignment is a relic from semicolon-less JS code. Since we're adding semicolons, it should go too.

File: lib/reactions/iteration.js (hunk 2 above)
Comment (devmanhinton) [inline lib/reactions/iteration.js]: Awesome, good catch

File: lib/rules.js [hunk 5]
```diff
@@ -59,16 +59,12 @@ rules.polymorphicValue = function(el) {
...
-  return el.children("." + name + "-").length
-}
-
-rules.collection = function(el) {
-  var collectionName = firstClassWhere(el, function(className) {
-    return this.hasSingularNamedChild(el, className)
-  }, this)
-  return collectionName ? undash(collectionName) : false
+rules.iteration = function(el) {
+  if(!(el.is("[data-each]"))) {
```
Comment (yaymukund) [inline lib/rules.js]: `return el.is('[data-each]')`

File: .gitignore [hunk 6]
```diff
@@ -1,4 +1,5 @@
 node_modules
+bower_components
```
Comment (yaymukund) [inline .gitignore:2]: Why this change?

File: docs/iteration.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+Backbone.Collection does not support attributes natively, but there are a number of
+options for extending collections to do so.  EndDash supports collection attributes
+as long as they are implemented as they are on Backbone.Model via the `get` method
+(which Backbone.Collection natively uses only for getting a model by id, not an
+attribute by name).  Typically collection attributes are used for metadata about
+the collection, such as total size (if the collection is paginated and this is
+different than length), as in the example below:
+
+```html
+<div class='users-' >
```
Comment (yaymukund) [inline docs/iteration.md:75]: Doublequotes for HTML attributes?

File: .gitignore (hunk 6 above)
Comment (devmanhinton) [inline .gitignore:2]: @bglusman

File: lib/reactions/model.js [hunk 8]
```diff
@@ -8,9 +8,14 @@ var ModelReaction = Reaction.extend({
 
   init: function(next, reload) {
     var newModel = this.get(this.modelName);
-    if(!(newModel instanceof Backbone.Model || newModel instanceof Backbone.Collection ||
-          this.isEndDashCompatible(newModel))) {
-      newModel = new Backbone.Model(newModel || {});
+
+    if(!(newModel instanceof Backbone.Model || newModel instanceof Backbone.Collection
```
Comment (yaymukund) [inline lib/reactions/model.js]: Do you need the first two checks anymore?

File: test/scope.js [hunk 9]
```diff
@@ -0,0 +1,40 @@
+require('./support/helper');
+
+var expect = require("expect.js")
+  , fs = require("fs")
+  , generateTemplate = require("./support/generate_template")
+  , Backbone = require('backbone')
+  , variableReaction = require('../lib/reactions/variable');
+
+  describe ("With a template that has a polymorphic key but no scope changes", function(){
```
Comment (tobowers) [inline test/scope.js]: would this be better as (pseudo code)

```
describe "scope", ()->
  describe "polymorphic key but no scope changes", ()->
    var markup = "blah"
    describe "model scoping reactions", ()->
      beforeEach ()->
        setup code from inside it

      it "should leave user", ()->

      it "should leave admin, ()->

```

File: test/support/templates/polymorphic.html [hunk 10]
```diff
@@ -1,4 +1,6 @@
-<ul class = "things- typePolymorphic-">
-  <li class = "thing- whenAwesome-">awesome</li>
-  <li class = "thing- whenCool-">cool</li>
+<ul class = "things-">
```
Comment (tobowers) [inline test/support/templates/polymorphic.html]: spacing consistency for class=""

File: test/iteration.js [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+      this.things = new Backbone.Collection([
+        new Backbone.Model({ type: "awesome" }),
+        new Backbone.Model({ type: "cool" })
+      ]);
+
+      this.markup = fs.readFileSync(__dirname + "/support/templates/polymorphic.html").toString()
+      this.template = generateTemplate({ things: this.things }, this.markup)
+    })
+
+    it("should change the item when the type changes", function() {
```
Comment (tobowers) [inline test/iteration.js:21]: I think I wrote this while fixing a bug, but it ain't good... waaaaayyyyy too long of a test.

File: lib/reactions/model.js (hunk 8 above)
Comment (devmanhinton) [inline lib/reactions/model.js]: Still need 

``` javascript
instanceof Backbone.Collection
```

But not

``` js
instanceof Backbone.Model
```

Refactoring to have checks be in function 

``` js
isEndDashCompatible
```

File: test/support/templates/polymorphic.html (hunk 10 above)
Comment (devmanhinton) [inline test/support/templates/polymorphic.html]: Addressing in next commit, thanks!

File: test/iteration.js (hunk 11 above)
Comment (devmanhinton) [inline test/iteration.js:21]: :) Cut down that spec quite a bit, INC (in next commit)

File: test/scope.js (hunk 9 above)
Comment (devmanhinton) [inline test/scope.js]: It is a bit clearer, INC

File: docs/iteration.md (hunk 7 above)
Comment (devmanhinton) [inline docs/iteration.md:75]: Punting this to the docs branch. Ie., when merge with that branch this will be overwritten

File: lib/rules.js (hunk 5 above)
Comment (devmanhinton) [inline lib/rules.js]: Addressed in commit

**Comment by devmanhinton** [comment]
@yaymukund  To address change from collection to iteration. It is a more semantic name. What that reaction does is iterate over the object in scope using its each method (and passing each to a template).

File: .gitignore (hunk 6 above)
Comment (bglusman) [inline .gitignore:2]: old version of bower installs to components directory, new one to bower_components looks like, so if you resintall with bower it's all in bower_componenets now and not git ignored, hence showing up as uncommitted files all the time.

**Comment by yaymukund** [comment]
@devmanhinton Not arguing that, but it makes it very difficult to review it. I'm just suggesting you send me a diff of just what has changed. If you can't do that, I can just review the whole file but it will take longer and you might have to fix stuff that you barely touched.

**Comment by bglusman** [comment]
@devmanhinton if we revert the name change and make that a separate PR after this is merged it will be cleaner... the diffs will reflect actual difference...  I should have insisted on that when we were working, even having the name change be a separate commit might do it...

File: .gitignore (hunk 6 above)
Comment (yaymukund) [inline .gitignore:2]: Wait, end-dash has a bower dependency? Where are these uncommitted files? How do they get there?

File: test/support/templates/collectionception.html [hunk 12]
```diff
@@ -1,10 +1,10 @@
-<ul class = "things-">
-  <li class = "thing-">
-    <ul class = "items-">
-      <li class = "item-">
-        <ul class = "objects-">
-          <li class = "object-">
-            <div class = "name-"></div>
+<ul class="things-">
```
Comment (yaymukund) [inline test/support/templates/collectionception.html:1]: Sigh, maybe not for this PR but we should use a real life example like `worlds -> continents -> countries -> states -> cities -> districts`... this test is cute but nonsensical.

Comment (devmanhinton) [inline test/support/templates/collectionception.html:1]: Sure, we can do that but like you say, not part of this PR

**Comment by devmanhinton** [comment]
@yaymukund Just did in the last 2 min

**Comment by yaymukund** [comment]
Will this work with regular arrays or do they have to be backbone collections?

**Comment by devmanhinton** [comment]
@yaymukund Yes & No. 

Depends as well if array has Backbone Models or literal models.

In general, EndDash will interpolate values of any model passed in (regardless of if it has a get method) but will only update on model/collection changes if that model/collection has get, set, on, & listenTo like Backbone objects.

So, in this case, with an array, it would interpolate the array's models into the DOM but would not update if an item was removed or added from, to, the array.

**Comment by bglusman** [comment]
@yaymukund @devmanhinton maybe I misunderstand your response and/or mukund's question, but I thought he meant the iteration working with them?  Internally we convert arrays into backbone collections if not already one, that's one of the changes we made to get this to work I think.

Oh, but good point on not updating if items added to the array, if it was created from an array...  wonder if that's a real issue we need to worry about or not...

**Comment by devmanhinton** [comment]
Ahhhh. Yes. Need to test again but yes. For the reason Brian mentioned (we wrap it in a Backbone.Collection)

File: docs/iteration.md [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+var usersCollection = new Backbone.Collection([user1, user2, user3]);
+```
+
+Iterate through the collection using data-each.
+(You must be in scope of the collection)
+
+```html
+<div data-each>
+  <div>
```
Comment (yaymukund) [inline docs/iteration.md:19]: Why do you need this inner div?

Comment (devmanhinton) [inline docs/iteration.md:19]: Anything inside data-each is converted to a template (partial reaction) which runs through parser.js. In parser.js we have a check to make sure each template has only 1 root. Thus, it must have a single root.

This can be changed but I would punt that to a another PR once this is merged.

File: lib/reactions/collection.js [hunk 14]
```diff
@@ -1,135 +1,116 @@
...
   addAll: function(models, index) {
-    index = index || 0
+    index = index || 0;
     if(_.isArray(models)) {
-      models = _(models)
-    }
-    models.each(function(item, index) {
-      this.add(item, index)
-    }, this)
+      models = _(models);
```
Comment (yaymukund) [inline lib/reactions/collection.js:13]: This is the wrong place to wrap `models`. Right now, I think we're doing:

``` js
init: function(model) {
  // We just store model blindly, no guarantees whether it's
  // backbone or non-backbone, underscore or
  // non-underscore.
  this.model = model;
},

addAll: function() {
  // Every time we call addAll, we have to do the dance where
  // we convert stuff into underscored stuff.
  if (_.isArray(model)) {
    model = _(model);
  }
},

// This isn't a great way to do it. At some point, maybe we have
// another method that works with the model as a collection. We'd
// have to do this whole dance again!
addSlightlyDifferent: function() {
  if (_.isArray(model)) {
    model = _(model);
  }
}
```

Instead, we should be doing that conversion once when we store the model. All other functions can just assume that it's been underscore-wrapped correctly.

Personally, I'd rather store it as a bare object and use underscore with the `_.each(models, function(model) {...})` syntax instead.

File: docs/iteration.md (hunk 13 above)
Comment (yaymukund) [inline docs/iteration.md:19]: I don't mind saving this for a subsequent PR, but this limitation should be documented so people aren't confused.

File: lib/reactions/collection.js [hunk 15]
```diff
@@ -1,135 +1,116 @@
...
-      , template = new Template(model, { stack: this.stack.slice(0) })
-      , children
+    var Template = this.getTemplateClass(model),
+        template = new Template(model, { stack: this.stack.slice(0) }),
+        children;
 
-    template.template.data("model", model)
+    template.el.data("model", model);
 
     if(typeof index === "undefined") {
```
Comment (yaymukund) [inline lib/reactions/collection.js:34]: This is pretty complex, low level logic here. Could it be in a `util` class?:

``` js
util.insertAt(this.templates, template, index);
util.insertElementAt(this.el, template.el, index);
```

File: docs/iteration.md (hunk 13 above)
Comment (devmanhinton) [inline docs/iteration.md:19]: agreed

File: lib/reactions/collection.js (hunk 15 above)
Comment (devmanhinton) [inline lib/reactions/collection.js:34]: could be, but that is a different PR, since this part of the code did not change in this PR.

File: lib/reactions/collection.js [hunk 16]
```diff
@@ -1,135 +1,116 @@
...
-  remove: function(index, model) {
+  remove: function(model) {
     var templateToRemove = _(this.templates).find(function(template) {
-      return template.template.data('model') === model;
-    })
-    //if (templateToRemove) {
-      templateToRemove.template.remove()
+      return template.el.data('model') === model;
+    });
+      templateToRemove.template.remove();
```
Comment (yaymukund) [inline lib/reactions/collection.js:51]: `.el.remove()`?

File: lib/reactions/collection.js (hunk 14 above)
Comment (devmanhinton) [inline lib/reactions/collection.js:13]: So,

``` js
  if (_.isArray(model)) {
    model = _(model);
  }
```

Is only hit when a collection is passed in a that top level without scoping. (otherwise it is wrapped by model.js)

I think a good solution to make this clearer is to wrap it in a Backbone.Collection in Template.js in bind.

If any of the above is unclear, feel free to ping me on campfire or in person

Comment (devmanhinton) [inline lib/reactions/collection.js:13]: See comment

```
//Hit only if array is passed in and iterated without scoping
```

File: lib/reactions/collection.js (hunk 16 above)
Comment (devmanhinton) [inline lib/reactions/collection.js:51]: yes. We did not go through and get ride of template.template references. That was not the goal of this PR.

Happy to change and push that though or do you want to save all those changes for another PR?

File: test/iteration.js (hunk 11 above)
Comment (yaymukund) [inline test/iteration.js:21]: Is this resolved? It still looks pretty long.

File: lib/reactions/collection.js (hunk 16 above)
Comment (yaymukund) [inline lib/reactions/collection.js:51]: If you're already touching a line, it's generally a good idea to also clean up stuff in the process because... why not?

If you want to save it for another PR, that's fine too. Whatever you think is best.

Comment (devmanhinton) [inline lib/reactions/collection.js:51]: I'll just make it another PR this time. But in the future I'll try to make incremental changes as I find them. Was just so focused on the code when writing was not thinking about this unrelated, albeit necessary, change.

File: lib/reactions/model.js [hunk 17]
```diff
@@ -8,9 +8,13 @@ var ModelReaction = Reaction.extend({
 
   init: function(next, reload) {
     var newModel = this.get(this.modelName);
```
Comment (yaymukund) [inline lib/reactions/model.js:10]: `this.get(this.modelName) || {};`

Again, we have this branching `if (_.isArray(foo)) foo = new Backbone.Foo(foo);` stuff. That should be handled in one place.

Comment (devmanhinton) [inline lib/reactions/model.js:10]: Please see the third discussion above this one.

Also, a lot of these are parts of the code that were the behavior of EndDash before this PR was started. I think that these are items are valid to consider and change but should be noted down as further PR.

I now have:

1) PR to replace template.template w/ template.el
2) PR to change when models are wrapped

Both of the above are valid and quick changes.

Comment (yaymukund) [inline lib/reactions/model.js:10]: Isn't the `new Backbone.Collection` stuff new to this branch?

File: lib/reactions/model.js [hunk 18]
```diff
@@ -22,6 +26,9 @@ var ModelReaction = Reaction.extend({
     if (!model) {
       return false;
     }
+    if (model instanceof Backbone.Collection) {
```
Comment (yaymukund) [inline lib/reactions/model.js:29]: Won't this fail for the same reason that `instanceof Backbone.Model` failed? (i.e. different instance of `Backbone` used by the app than by EndDash internally)

Comment (devmanhinton) [inline lib/reactions/model.js:29]: Yes, that is a good point. Perhaps define a helper that checks to see if .each is defined on the object?

File: lib/reactions/model.js (hunk 17 above)
Comment (devmanhinton) [inline lib/reactions/model.js:10]: Not totally, we did change it since it was missing an edge case with new collection iteration. So we changed it, slightly.

File: lib/reactions/model.js (hunk 18 above)
Comment (devmanhinton) [inline lib/reactions/model.js:29]: How would you feel about not wrapping objects at all. Assume that if you pass in a model that has the correct interface. And if it is a collection that it can be iterated through?

Comment (yaymukund) [inline lib/reactions/model.js:29]: I feel like the correct solution is to use an existing backbone if one is defined when EndDash starts up.

I don't think that belongs in this PR, though, so don't worry about it. I'll open up a new issue.

Comment (devmanhinton) [inline lib/reactions/model.js:29]: Great. Yes. It will be nice once that issue is resolved :D :dancer:

File: .gitignore (hunk 6 above)
Comment (yaymukund) [inline .gitignore:2]: Bump!

File: test/iteration.js (hunk 11 above)
Comment (devmanhinton) [inline test/iteration.js:21]: I believe the test that was 'too long' was 'should change the item when the type changes'. This was 10 lines. Now it is 6..

File: .gitignore (hunk 6 above)
Comment (bglusman) [inline .gitignore:2]: run bower install in end-dash repo and see for yourself?  that's what the bower.json file is for, that's where inflection library comes from, and jquery and underscore.

Comment (bglusman) [inline .gitignore:2]: PS - doing Bump without mentioning anyone by name makes it harder to notice :-)

Comment (bglusman) [inline .gitignore:2]: which, taking my own advice, means I should @yaymukund you in a reply :-)
